Re-attempt support for aux kernel evaluation of lower-d vars

lindsayad opened the pull request:
This has the revert of the revert of #26593 

Additionally, in order to get all tests to pass in parallel and distributed recipes, this also now has significant robustness improvements for relationship managers and for some other framework objects; the improvements for non-relationship manager objects make them robust against late remote element deletion. Late remote element deletion was newly triggered in a lot of inputs due to the addition of the `GhostLowerDElems` RM to aux kernels.

lindsayad committed:
Make sure to ghost lower-d elements for aux kernels that require them

Refs parallel failures from #26593

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/2021720/) on 0aeffb5 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/26625/site)

This comment will be updated on new commits.

lindsayad committed:
Add test mesh generator that always deletes its remote elements

This allows us to avoid regolding Fande's test

lindsayad committed:
Remove unnecessary check for MeshBase object

lindsayad committed:
Make ElementalVariableValue robust against late element deletion

lindsayad committed:
Switch the `_moose_mesh` to be the displaced mesh when appropriate

If the libMesh mesh is the displaced one, then our MooseMesh should
be as well to help prevent bugs in user code

lindsayad committed:
Don't find neighbors for lower-d elements on internal faces

The selection of a neighbor is ambiguous in this case and in my
experience can lead to failing libmesh assertions down the road

lindsayad committed:
Delay initialization of the point locator for `PiecewiseConstantFromCSV`

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/2023722/) on 40275bf wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>1c3505</th>
      <th colspan=4>#26625 40275b</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>85.23%</td>
      <td>85.24%</td>
      <td>+0.01%</td>
      <td>100.00%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>99135</td>
      <td>99152</td>
      <td>+17</td>
      <td>19</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>17173</td>
      <td>17169</td>
      <td>-4</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/26625/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/26625/coverage/framework)


## Modules coverage

Coverage did not change

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/26625/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/26625/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/26625/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/26625/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/26625/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/26625/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/26625/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/26625/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/26625/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/26625/coverage/geochemistry)
- [`heat_transfer`](https://mooseframework.inl.gov/docs/PRs/26625/coverage/heat_transfer)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/26625/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/26625/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/26625/coverage/navier_stokes)
- [`optimization`](https://mooseframework.inl.gov/docs/PRs/26625/coverage/optimization)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/26625/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/26625/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/26625/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/26625/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/26625/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/26625/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/26625/coverage/richards)
- [`scalar_transport`](https://mooseframework.inl.gov/docs/PRs/26625/coverage/scalar_transport)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/26625/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/26625/coverage/stochastic_tools)
- [`tensor_mechanics`](https://mooseframework.inl.gov/docs/PRs/26625/coverage/tensor_mechanics)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/26625/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/26625/coverage/xfem)

</details>



This comment will be updated on new commits.

lindsayad committed:
Explicitly initialize RMs with a MooseMesh

This is a lot better than guessing what moose mesh they might be
using. Moreover, a user can create arbitrary meshes that are not
the two we know about in the framework

lindsayad committed:
Revert "Revert "Support aux kernel evaluation of lower-d vars""

This reverts commit d529c49b630fbe2ab272bc0206038f7f608b766b.

moosebuild commented:
Job [Documentation internal apps](https://civet.inl.gov/job/2026090/) on 02bfdc9 : invalidated by @lindsayad

moosebuild commented:
Job [Internal app tests](https://civet.inl.gov/job/2026121/) on 02bfdc9 : invalidated by @lindsayad

moosebuild commented:
Job [OpenMPI](https://civet.inl.gov/job/2026129/) on 02bfdc9 : invalidated by @lindsayad

lindsayad committed:
Add error check in MooseMesh::buildTypedMesh

This should stop erroneous toggling of the moose mesh's parallel
type

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/2035601/) on b9b05f0 wanted to post the following:


The following coverage requirement(s) failed:
- Failed to generate `framework` coverage rate (required: 81.0%)
- Failed to generate `tensor_mechanics` coverage rate (required: 84.0%)

GiudGiud commented on test/src/meshgenerators/AlwaysDeleteRemotesGeneratedMeshGenerator.C:
```diff
@@ -0,0 +1,322 @@
+
+AlwaysDeleteRemotesGeneratedMeshGenerator::AlwaysDeleteRemotesGeneratedMeshGenerator(
+    const InputParameters & parameters)
+  : MeshGenerator(parameters),
```
why not start from a GeneratedMG base class?

GiudGiud commented on framework/src/postprocessors/ElementalVariableValue.C:
```diff
@@ -39,6 +38,16 @@ ElementalVariableValue::ElementalVariableValue(const InputParameters & parameter
+void
+ElementalVariableValue::initialSetup()
+{
+  // Do this in initialSetup instead of the constructor because elements may be deleted after this
+  // object is constructed
+  _element = _mesh.queryElemPtr(getParam<unsigned int>("elementid"));
```
this seems problematic

shouldnt we stop deleting elements after the mesh is done? because otherwise the ids after --mesh-only wont match the ids during the simulation

lindsayad replied:
We have had late geometric ghosting since the inception of relationship managers. This is to allow algebraic objects to give us more information about what needs to be ghosted after their construction. For instance we do this with FV kernels a lot; we inspect some information in the FV object constructor and then decide whether to expand the number of ghost layers. So the concept has always been around. Just some objects have been able to get away with not respecting that possibility until now

lindsayad replied:
> because otherwise the ids after --mesh-only wont match the ids during the simulation

I don't really see a problem if the element ids present after a `--mesh-only` are a superset of what's available during the simulation. If there is a problem, then that means somebody hasn't properly advertised their ghosting need

GiudGiud replied:
I guess the problem is not new. The ids can be re-numbered between --mesh-only and a regular run

GiudGiud replied:
People should just never use Ids I guess. It s too risky with the potential renumbering.

lindsayad replied:
Yes relying on ids was a bad idea even before `DistributedMesh` existed

lindsayad committed:
Have test generator inherit directly from GeneratedMeshGenerator

This allows not re-writing the class definition, validParams, nor
the constructor. Unfortunately we still have to re-write `generate`
since we have to insert our remote element deletion line

GiudGiud commented on test/src/meshgenerators/AlwaysDeleteRemotesGeneratedMeshGenerator.C:
```diff
@@ -27,89 +27,13 @@ registerMooseObject("MooseTestApp", AlwaysDeleteRemotesGeneratedMeshGenerator);
-    mooseError("Cannot apply both Gauss-Lobatto mesh grading and biasing at the same time.");
 }
 
 std::unique_ptr<MeshBase>
```
could you just call generate from the GMG too and add the one line after?

lindsayad replied:
You have to insert `allow_remote_element_removal(false)` in between `builMeshBaseObject` and `build_square/cube`

lindsayad replied:
Ok good for you for pushing on this more. Most recent commit I was able to reuse a lot more

GiudGiud replied:
nice

GiudGiud reviewed (commented):
just the one comment then I m good on this

lindsayad committed:
Duplicate less code in AlwaysDeleteRemotesGeneratedMeshGenerator

lindsayad commented:
failures are unrelated. Thanks for the reviews!

lindsayad commented:
for sqa, this was the last testing report https://civet.inl.gov/event/165852/ before I dumbly closed and reopened the issue
